### PR TITLE
Add ManageSieve support

### LINKED ISSUE #53: Unable to activate a Sieve script via ManageSieve
Hi, I'm migrating my mail server from Debian to nixos-mailserver and it mostly works great! Thanks for the project. :)

The only problem I've faced so far is broken activation of Sieve scripts. I was successfully able to upload a Sieve script with KMail, but trying to activate the scripts results in this error:

> managesieve(user@example.com): Error: sieve: file script: Failed to activate Sieve script: symlink(/var/sieve/user@example.com.sieve/default.sieve, /var/empty/.dovecot.sieve) failed: Operation not permitted

I'm using nixos-mailserver-2.0.4 on NixOS 17.09.

Thanks!

### PR DESCRIPTION
**Pull request by erosennin**
Fixes #53.

### CODE REVIEW

File: mail-server/dovecot.nix [hunk 1]
```diff
@@ -109,7 +109,9 @@ in
         }
 
         plugin {
-          sieve = file:/var/sieve/%u.sieve
+          sieve = file:/var/sieve/%u/scripts;active=/var/sieve/%u/active.sieve
```
Comment (r-raymond) [inline mail-server/dovecot.nix]: Are these folders created automatically by Dovecot's managesieve?

File: mail-server/users.nix [hunk 2]
```diff
@@ -55,16 +55,16 @@ let
     # is null, remove the file.
     ${lib.concatMapStringsSep "\n" ({ name, sieveScript }:
       if lib.isString sieveScript then ''
-        cat << EOF > "/var/sieve/${name}.sieve"
+        cat << EOF > "/var/sieve/${name}/default.sieve"
```
Comment (r-raymond) [inline mail-server/users.nix]: I think we need a
`mkdir /var/sieve/${name}`
here

File: mail-server/users.nix [hunk 3]
```diff
@@ -55,16 +55,16 @@ let
...
         ${sieveScript}
         EOF
-        chown "${name}:${vmailGroupName}" "/var/sieve/${name}.sieve"
+        chown "${name}:${vmailGroupName}" "/var/sieve/${name}/default.sieve"
       '' else ''
-        if (test -f "/var/sieve/${name}.sieve"); then
-          rm "/var/sieve/${name}.sieve"
+        if (test -f "/var/sieve/${name}/default.sieve"); then
+          rm "/var/sieve/${name}/default.sieve"
```
Comment (r-raymond) [inline mail-server/users.nix]: I guess here we can simply remove the folder now

File: default.nix [hunk 4]
```diff
@@ -280,6 +280,15 @@ in
       '';
     };
 
+    enableManageSieve = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable ManageSieve, setting this option to true will open
+        port 4190 in the firewall.
```
Comment (r-raymond) [inline default.nix]: Maybe a quick explanation that this allows to manage sieve scripts via thunderbird etc.

File: mail-server/dovecot.nix [hunk 5]
```diff
@@ -41,7 +41,7 @@ in
       protocols = [ "sieve" ];
 
       sieveScripts = {
-        before = builtins.toFile "spam.sieve" ''
+        after = builtins.toFile "spam.sieve" ''
```
Comment (r-raymond) [inline mail-server/dovecot.nix]: Can you elaborate on what the interaction of this and the line

```
sieve_default = file:/var/sieve/%u/default.sieve
sieve_default_name = default
```
is?

File: default.nix (hunk 4 above)
Comment (erosennin) [reply default.nix]: OK, done.

File: mail-server/dovecot.nix (hunk 5 above)
Comment (erosennin) [reply mail-server/dovecot.nix]: The `sieve_default` script is executed first. If a matching rule is found in this script, the corresponding action is applied and `spam.sieve` is not executed (unless the action is `keep`). If no role matches, the Sieve interpreter continues to the `spam.sieve` script.

File: mail-server/dovecot.nix (hunk 1 above)
Comment (erosennin) [reply mail-server/dovecot.nix]: Yes.

File: mail-server/users.nix (hunk 3 above)
Comment (erosennin) [reply mail-server/users.nix]: The folder may still contain user's scripts uploaded via ManageSieve, so let's just keep it.

File: mail-server/users.nix (hunk 2 above)
Comment (erosennin) [reply mail-server/users.nix]: Right, fixed, thanks.

**Comment by r-raymond** [comment]
Awesome, I can't find any more problems. Thanks for the changes!

Did you test if using managesieve works? I have no idea how to write a test for that :)

@eqyiel: in the newest version, the username's are not real accounts anymore. Is it save to set vmail as the owner of the scripts?

**Comment by erosennin** [comment]
Works for me in KMail and in Thunderbird with the Sieve plugin.

We could use some console client for testing, like [sieve-connect](https://github.com/philpennock/sieve-connect), but it's not in Nix packages and packaging it seems a bit too much of a pain right now. :)

**Comment by eqyiel** [comment]
@r-raymond are you talking about this commit? https://github.com/r-raymond/nixos-mailserver/commit/2d0648e0f4c57e8a57ad34bbd8ea514fcf18c543

IIRC, in this case the owner of the scripts has to be vmail otherwise the virtual users won't be able to modify them.

@erosennin I'm using this right now and it seems fine to me, thanks for looking into it!

**Comment by r-raymond** [comment]
Yes, that is the commit I'm worried about. Shouldn't that break
```
chown "${name}:${vmailGroupName}" "/var/sieve/${name}"
```
?

**Comment by eqyiel** [comment]
@r-raymond I see what you mean - in that case this should be changed to `chown "${vmailUserName}:${vmailGroupName}" "/var/sieve/${name}"` in 2.1, AFAICT it will be sufficient for dovecot to have write permissions to the file after the switch to passwd-file users (but I'll check it out if you want to rebase that onto master after this lands) 👍

**Comment by r-raymond** [comment]
@eqyiel Ok, then I'll merge this but I'll fix the chowns before I'll post a new release.

Thanks everyone!

**Pull request merged**